com.unity.shadergraph/Editor/GraphUI/EditorCommon/CommandStateObserver/NodeUICommands.cs

```diff
@@ -0,0 +1,79 @@
+using UnityEditor.GraphToolsFoundation.Overdrive;
+using UnityEditor.ShaderGraph.Registry;
+using UnityEditor.ShaderGraph.Registry.Types;
+using UnityEngine;
+using UnityEngine.GraphToolsFoundation.CommandStateObserver;
+
+namespace UnityEditor.ShaderGraph.GraphUI
+{
+    class SetGraphTypeValueCommand : UndoableCommand
```

Comment:
Maybe. The crux of this command is calling `SetPortField` on a node writer to set some values. If that and notifying the preview manager are useful for setting blackboard properties, then it could be reusable.

But I'm guessing it probably won't end up being too helpful because of the focus on writing to a port. At best maybe some of the logic (like translating a `float[]` into a `Matrix4x4` for the material property block) could be factored out into a utility function.

`UpdateConstantValueCommand` is definitely similar, but goes through GTF's `IConstant` which isn't useful for our static GraphDelta ports.

Reply:
Hmm we'll have to circle back to this when we have blackboard properties in, because yeah the core of what we do for updating blackboard properties (as well as any port value really) is what you mention there, setting the value on the node writer and then notifying the preview manager, and then the GTF data model gets updated from the GraphStorage/CLDS level, so we're kinda bypassing the GTF data model entirely, as far as I understand it right now